### PR TITLE
In MBEDTLS_ARIA_C config, implemeting 16 bits swap.

### PR DESCRIPTION
**Pull request by devnexen**
Signed-off-by: David Carlier <devnexen@gmail.com>

## Status
**READY**

## Requires Backporting
NO  

## Migrations
If there is any API change, what's the incentive and logic for it.

NO

### CODE REVIEW

File: library/aria.c [hunk 1]
```diff
@@ -112,8 +112,12 @@ static inline uint32_t aria_p1( uint32_t x )
 #endif /* arm */
 #if defined(__GNUC__) && \
     defined(__i386__) || defined(__amd64__) || defined( __x86_64__)
-/* I couldn't find an Intel equivalent of rev16, so two instructions */
-#define ARIA_P1(x) ARIA_P2( ARIA_P3( x ) )
+static inline uint32_t aria_p1( uint32_t x )
+{
+    uint32_t r;
+    __asm( "xchgb %b0, %h0" : "=q"(r) : "0"(x) );
```
Comment (mpg) [inline library/aria.c:118]: I'm only moderately familiar with x86 assembly, and it is unclear to me what's going on here, so I think a comment would be helpful.

File: library/aria.c [hunk 2]
```diff
@@ -112,8 +112,12 @@ static inline uint32_t aria_p1( uint32_t x )
...
+    return( r );
+}
```
Comment (mpg) [inline library/aria.c:120]: You need to also `#define ARIA_P1 aria_p1` here, or the default version below is going to be used. (Clang helpful gives a warning about `aria_p1` being unused, which GCC doesn't.)

Comment (mpg) [reply library/aria.c:120]: Note: after adding this line (and enabling ARIA in `config.h`), trying to build the library give the following error with GCC:
```
aria.c: In function ‘aria_fo_xor’:
aria.c:388:1: error: extended registers have no high halves
```
Clang also complains:
```
aria.c:118:12: error: invalid operand for instruction                                                                                                                                                         
    __asm( "xchgb %b0, %h0" : "=q"(r) : "0"(x) );                                                                                                                                                             
           ^                                                                                                                                                                                                  
<inline asm>:1:14: note: instantiated into assembly here                                                                                                                                                      
        xchgb %bpl, %bp                                                                                                                                                                                       
                    ^~~ 
```

**Review by mpg — CHANGES_REQUESTED** [review]
Thanks for your contribution! This would indeed be a nice improvement, however there are a couple of issues to be resolved.

**Comment by mpg** [comment]
I tried the proposed assembly in a separate file in order to try understanding how it works while avoiding the register selection issues that seem to be happening when inlined in other functions in `aria.c`, but it doesn't seem to be giving the expected result:
```c
#include <stdint.h>
#include <stdio.h>

uint32_t aria_p1( uint32_t x )
{
    uint32_t r;
    __asm( "xchgb %b0, %h0" : "=q"(r) : "0"(x) );
    return( r );
}

#define ARIA_P1(x) ((((x) >> 8) & 0x00FF00FF) ^ (((x) & 0x00FF00FF) << 8))

int main(void) {
    uint32_t x = 0xaabbccdd;
    printf("%08x -> %08x (ref: %08x)\n", x, aria_p1(x), ARIA_P1(x));
}
```
Compiling and executing:
```
% gcc -Wall -Wextra -O2 xchgb.c -o xchgb && ./xchgb
aabbccdd -> aabbddcc (ref: bbaaddcc)
```
It appears that the top bytes have not been swapped as they should. It's unclear to me whether there's a way to fix this without going back to two instructions.

**Comment by devnexen** [comment]
You re right I can t come up without two instructions here.

**Comment by mpg** [comment]
Thanks for trying anyway, this would have been a nice improvement!

**Pull request closed without merging**